### PR TITLE
Private claims implementation

### PR DESCRIPTION
**Pull request by skroczek**
This PullRequest should implement private claims as described in RFC7519 in section 4.2. According to #1120 , this is still missing, but is wanted.
This implementation is based on that of scopes, which is basically also only a (registered) claim.
I don't think the implementation is finished yet. There are a few questions open ~~and the testing is still missing completely~~. But the code works and can already be used for evaluation.
One open question is whether the claims should be checked for registered claims before the token is generated. If so, we would have to think about how to react in case of a violation. Will the claim simply be ignored, or should an exception be thrown?
~~About the testing: I have already looked at the test suite, but I'm not quite sure where and how to put it there.~~ A suggestion would be very welcome.

In general I would be very happy about feedback.

### CODE REVIEW

File: src/Entities/ClaimEntityInterface.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * @author      Sebastian Kroczek <me@xbug.de>
+ * @copyright   Copyright (c) Sebastian Kroczek
+ * @license     http://mit-license.org/
+ *
+ * @link        https://github.com/thephpleague/oauth2-server
+ */
+
+namespace League\OAuth2\Server\Entities;
+
+use JsonSerializable;
```
Comment (Pchelolo) [inline src/Entities/ClaimEntityInterface.php:12]: I couldn't find a reason for this to extend JsonSerializable? I couldn't find any.

File: src/Entities/ClaimEntityInterface.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+interface ClaimEntityInterface extends JsonSerializable
```
Comment (Pchelolo) [inline src/Entities/ClaimEntityInterface.php:14]: It would be neat to provide a ClaimEntityTrait, similar to TokenEntityTrait to simplify implementation.

File: tests/Stubs/ClaimEntity.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        return $this->name;
+    }
+
+    public function getValue()
+    {
+        return $this->value;
+    }
+
+    public function jsonSerialize()
```
Comment (Pchelolo) [inline tests/Stubs/ClaimEntity.php:29]: If it's needed (see comment above), $this->value being mixed is not guaranteed to be jsonSerializable itself

File: src/Repositories/ClaimRepositoryInterface.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /**
+     * Returns claims
+     *
+     *
+     * @return ClaimEntityInterface[]
+     */
+    public function getClaims(
+        array $claims,
+        $grantType,
```
Comment (Pchelolo) [inline src/Repositories/ClaimRepositoryInterface.php:28]: grantType will be a string, can typehint

File: src/Repositories/ClaimRepositoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Claim repository interface.
+ */
+interface ClaimRepositoryInterface extends RepositoryInterface
+{
+    /**
+     * Returns claims
+     *
+     *
+     * @return ClaimEntityInterface[]
```
Comment (Pchelolo) [inline src/Repositories/ClaimRepositoryInterface.php:24]: params docs are missing

File: src/Repositories/ClaimRepositoryInterface.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    public function getClaims(
+        array $claims,
```
Comment (Pchelolo) [inline src/Repositories/ClaimRepositoryInterface.php:27]: What is the idea behind passing an array $claims in? AFAIK it's not needed

File: tests/Grant/AbstractGrantTest.php [hunk 7]
```diff
@@ -360,7 +361,8 @@ public function testIssueAccessToken()
             new DateInterval('PT1H'),
             new ClientEntity(),
             123,
-            [new ScopeEntity()]
+            [new ScopeEntity()],
+            [new ClaimEntity('private', 'claim')]
```
Comment (Pchelolo) [inline tests/Grant/AbstractGrantTest.php:365]: Would be good to have tests for non-string claim values as well

File: tests/Grant/ClientCredentialsGrantTest.php [hunk 8]
```diff
@@ -43,10 +44,14 @@ public function testRespondToRequest()
         $scopeRepositoryMock->method('getScopeEntityByIdentifier')->willReturn($scope);
         $scopeRepositoryMock->method('finalizeScopes')->willReturnArgument(0);
 
+        $claimRepositoryMock = $this->getMockBuilder(ClaimRepositoryInterface::class)->getMock();
+        $claimRepositoryMock->method('getClaims')->willReturn([]);
```
Comment (Pchelolo) [inline tests/Grant/ClientCredentialsGrantTest.php:48]: Would be good to have a test where $claimRepo is returning something

File: tests/ResponseTypes/BearerResponseTypeTest.php [hunk 9]
```diff
@@ -32,11 +33,14 @@ public function testGenerateHttpResponse()
...
         $scope->setIdentifier('basic');
 
+        $claim = new ClaimEntity('_private', 'claim');
+
         $accessToken = new AccessTokenEntity();
         $accessToken->setIdentifier('abcdef');
         $accessToken->setExpiryDateTime((new DateTimeImmutable())->add(new DateInterval('PT1H')));
         $accessToken->setClient($client);
         $accessToken->addScope($scope);
+        $accessToken->addClaim($claim);
```
Comment (Pchelolo) [inline tests/ResponseTypes/BearerResponseTypeTest.php:43]: Would be nice to have an assertion downstream that the claim has been set

File: tests/Stubs/ClaimEntity.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+
+namespace LeagueTests\Stubs;
+
+use League\OAuth2\Server\Entities\ClaimEntityInterface;
+
+class ClaimEntity implements ClaimEntityInterface
```
Comment (Pchelolo) [inline tests/Stubs/ClaimEntity.php:8]: In case you had a ClaimEntityTrait this would be much simpler, like other stub objects

File: tests/Grant/AbstractGrantTest.php (hunk 7 above)
Comment (Pchelolo) [reply tests/Grant/AbstractGrantTest.php:365]: Would be good to have an assertion verifying the claim was set.

**Comment by Pchelolo** [comment]
I have left a few comments inline for the things that popped out. We have not yet fully tested this implementation, will report the results as we go, but this looks great overall. 

I have been looking into standards around this, and haven't found much: OAuth spec doesn't specify what would be used as a token, thus it doesn't contain any requirements or restrictions regarding the claims we can add into JWT. It seems logical that the library itself wouldn't impose any additional restrictions, so in my opinion the ClaimEntityInterface is rightfully very generic. 

I've surveyed a few other popular oauth2 server frameworks that support to adding custom claims within OAuth:
- Spring Security: allows to register a `TokenEnhancer` interface with the authentication manager. The enhancer receives the token object and a context-like Authentication object and can set additional information to the token object. [Example](https://sultanov.dev/blog/custom-jwt-claims-in-spring-security-oauth/). That is much more generic then what's implemented in the pull request, however this library encapsulates more logic than spring does, so I think following more generic approach taken by spring would be an overkill here. 
- nodejs [oauth2-server](https://oauth2-server.readthedocs.io/en/latest/index.html) - leaves generating the access token entirely to the library client, does not care about the access token semantics, treating it as an opaque string. Not viable approach here either
- python [OAuthLib](https://oauthlib.readthedocs.io/en/latest/index.html) - allows to provide a custom token generator, treating the token as an opaque string. alternatively has JWT-based implementation, that allows setting custom claims and is fairly similar to what's implemented here.

To sum up the approach taken in this PR seems reasonable to me, is on par with the capabilities of other libraries and does not violate any existing standards or RFCs as far as I could see.

File: src/Entities/ClaimEntityInterface.php (hunk 1 above)
Comment (skroczek) [reply src/Entities/ClaimEntityInterface.php:12]: You are right, this is really not needed since the builder takes care of it. I guess I just took it from the scope implementation. Nice catch.

File: src/Repositories/ClaimRepositoryInterface.php (hunk 4 above)
Comment (skroczek) [reply src/Repositories/ClaimRepositoryInterface.php:28]: Right, I'll add that.

File: src/Repositories/ClaimRepositoryInterface.php (hunk 6 above)
Comment (skroczek) [reply src/Repositories/ClaimRepositoryInterface.php:27]: Good point, again the Scope implementation was used as a prototype. However, it seems that I did not use the right name, it should be finalizeClaims. Also the comment should be adjusted accordingly. Anyway, the point of the method is that the claims can be validated and optionally more can be added or removed. This is especially useful when working with listeners. If there are no further concerns, I would simply rename the method and add an usefull comment accordingly.

**Comment by skroczek** [comment]
Thank you very much for the close look and for the feedback. I have already answered to a few points. The suggestions all make sense to me, so I will implement them over the next days.

**Comment by Pchelolo** [comment]
@skroczek do you need any assistance implementing/testing your changes?

File: tests/Grant/AbstractGrantTest.php (hunk 7 above)
Comment (skroczek) [reply tests/Grant/AbstractGrantTest.php:365]: This took some time, but this is not the right place to test values of the AccessToken. In this test the AccessToken is returned by a mock as a fixed return value. Since the claim parameter for the method AbstractGrant::issueAccessToken is optional, I will replace it with an empty array to match the value of the returned AccessToken.

File: tests/ResponseTypes/BearerResponseTypeTest.php (hunk 9 above)
Comment (skroczek) [reply tests/ResponseTypes/BearerResponseTypeTest.php:43]: As in the ClientCredentialsGrantTest, the AccessToken is returned by a mock. Therefore it is not possible to test whether claims are set here, at least not without fundamentally rewriting the test.

**Comment by skroczek** [comment]
@Pchelolo It took me a while to find time to take care of it. But I think I've implemented all your suggestions except for the two tests. Thanks again for your suggestions and your patience. If you notice anything else, feel free to bring it up.

**Comment by Pchelolo** [comment]
Thanks a lot and sorry that some of the suggestions didn't make sense. The CI seem to be failing. We will test the new version out and I'll share the results

File: src/Repositories/AccessTokenRepositoryInterface.php [hunk 11]
```diff
@@ -24,11 +25,17 @@ interface AccessTokenRepositoryInterface extends RepositoryInterface
...
+     * @param ClaimEntityInterface[] $claims
      *
      * @return AccessTokenEntityInterface
      */
-    public function getNewToken(ClientEntityInterface $clientEntity, array $scopes, $userIdentifier = null);
+    public function getNewToken(
+        ClientEntityInterface $clientEntity,
+        array $scopes,
+        $userIdentifier = null,
+        array $claims = []
```
Comment (Pchelolo) [inline src/Repositories/AccessTokenRepositoryInterface.php:37]: Adding this new parameter to the interface method makes it incompatible with older library versions, php throws a fatal error when using old code with the new library due to incompatible method declaration, and this is the only backwards-incompatible change I could find. 

I'm not sure what's the best way to mitigate the problem. Some options:
1. Not changing the interface and making the client code responsible for adding the claims via addClaim(). In the next major version of the library the interface can be updated to have this extra parameter
2. Making this change and releasing a new major version of the library with this change.

I'm not a maintainer of the library, so asking @Sephster for judgement.

**Comment by Pchelolo** [comment]
We've tested this implementation more and it seems to work well. I've left a comment about backwards compatibility inline, otherwise this LGTM. I'm not a maintainer of the library, so I would ask @Sephster to kindly look at it.

File: src/Repositories/AccessTokenRepositoryInterface.php (hunk 11 above)
Comment (skroczek) [reply src/Repositories/AccessTokenRepositoryInterface.php:37]: Good point, my intention was actually to avoid backward incompatibilities. However, I agree with you that it boils down to the possibilities you mentioned. Basically it is up to @Sephster to decide if it is OK for him to create a new major version for the pull request or if I should adapt this change to be able to use it in a minor version. For me both solutions are fine, but I think that a minor version could be released sooner.

Comment (Pchelolo) [reply src/Repositories/AccessTokenRepositoryInterface.php:37]: ping @Sephster

Comment (Sephster) [reply src/Repositories/AccessTokenRepositoryInterface.php:37]: Thank you both for your work on this so far. I have a bit of time off work so will be aiming to review this myself in due course but the work you have both done has been invaluable.

I'd prefer not to release a major version for this change so I think it would be preferable to avoid any BC changes. If we want to keep the implementation as is, I'd merge it into next major branch but I don't think it would be released for some time to come yet.

I would lean towards going with the `addClaim()` method proposal instead if possible. Thank you

Comment (skroczek) [reply src/Repositories/AccessTokenRepositoryInterface.php]: I have adapted the arguments of the method, but I noticed something.Because the pull-request also changes interfaces, I wonder if it is even possible not to have back incompatibilities. This is not directly noticeable, because the interfaces are implemented by traits, which I have adapted accordingly and therefore no errors occur. However, these do not have to be used, which means that you might have to change your own implementation.
Depending on how it is handled in this project, we might have to wait for a major release. Although I would love to see the pull request accepted soon.

File: src/Entities/Traits/AccessTokenTrait.php [hunk 12]
```diff
@@ -42,13 +43,20 @@ public function setPrivateKey(CryptKey $privateKey)
...
             ->expiresAt($this->getExpiryDateTime()->getTimestamp())
-            ->relatedTo((string) $this->getUserIdentifier())
+            ->relatedTo((string) $this->getUserIdentifier());
+
+        foreach ($this->getClaims() as $claim) {
+            $builder->withClaim($claim->getName(), $claim->getValue());
+        }
+
+        return $builder
+            // Set scope claim late to prevent it from being overridden.
```
Comment (Sephster) [inline src/Entities/Traits/AccessTokenTrait.php:59]: Is this correct? I think an exception is thrown if we try to register the same claim twice.

**Review by Sephster — COMMENTED** [review]
I've added in some minor comments. If these could be addressed I'm happy to merge this in. 

Unfortunately this will be a break change release due to the changes to the interfaces to add the `getClaims`, `addClaim`, and `setClaimRepository` methods so will need to be in a major release.

If you are happy for me to do so, I will merge this into the 9.0.0-WIP branch instead of the master branch. Thanks for all of your work on this and do let me know if you have any comments

File: src/Entities/Traits/AccessTokenTrait.php (hunk 12 above)
Comment (skroczek) [reply src/Entities/Traits/AccessTokenTrait.php:59]: Yeah, I think that's correct. I took another look at the code of the library, and the builder does not check if a claim has already been added. At least I couldn't find anything.

**Comment by skroczek** [comment]
I already expected, after the last review, that the pull request will not make it into the current master branch. I noticed the breaking changes quite late. I have implemented your suggested changes so far. Anyway, I agree of course, if it is merged into the 9.0.0-WIP branch.  

Currently there seems to be a problem with styleci, but unfortunately I couldn't find out why the check fails. As far as I have seen, no changes have been made that could cause the check to fail. However, this is also the first time I am working with this service.

I will be poorly reachable in the next two weeks, so it may take some time before I can respond.

**Comment by Sephster** [comment]
Thanks @skroczek for the changes. If you are happy to, I can pick up any remaining issues to push this through to the 9.x branch? Will aim to do that tomorrow

**Comment by skroczek** [comment]
Hi @Sephster, if you could take care of the rest of these issues, I would be very grateful. Thank you very much for your support.

**Comment by Pchelolo** [comment]
Hello, anything we can do to help getting this in quicker?

File: src/Entities/Traits/AccessTokenTrait.php (hunk 12 above)
Comment (Sephster) [reply src/Entities/Traits/AccessTokenTrait.php:59]: Cool - thanks for checking.

File: src/Entities/ClaimEntityInterface.php (hunk 2 above)
Comment (Sephster) [reply src/Entities/ClaimEntityInterface.php]: Added so resolving comment.

File: tests/Stubs/ClaimEntity.php (hunk 3 above)
Comment (Sephster) [reply tests/Stubs/ClaimEntity.php]: Function has been removed. Marking as resolved.

File: src/Repositories/ClaimRepositoryInterface.php (hunk 4 above)
Comment (Sephster) [reply src/Repositories/ClaimRepositoryInterface.php]: Added. Marking as resolved.

File: src/Repositories/ClaimRepositoryInterface.php (hunk 6 above)
Comment (Sephster) [reply src/Repositories/ClaimRepositoryInterface.php]: Removed so marking as resolved.

File: src/Entities/TokenInterface.php [hunk 13]
```diff
@@ -82,4 +82,18 @@ public function addScope(ScopeEntityInterface $scope);
      * @return ScopeEntityInterface[]
      */
     public function getScopes();
+
+    /**
+     * Associate a claim with the token.
```
Comment (Sephster) [inline src/Entities/TokenInterface.php:87]: I'm going to move these functions out of this interface. Claims aren't part of the OAuth spec. To keep this library as flexible as possible, we've kept the JWT specific implementation out of the interfaces so that if you don't want to use it, you don't have to. 

These functions would be better placed in the trait where we keep a similar JWT specific function, `convertToJWT`.  I will move this there along with the `getClaims()` function. Thanks

Comment (skroczek) [reply src/Entities/TokenInterface.php]: I am totally fine with that. Thank you very much.

**Comment by garethellis36** [comment]
Hello folks. Is this PR likely to be merged soon? I think it would help me with a problem I am facing.

**Comment by Sephster** [comment]
No timescale on this I'm afraid. It will be done as soon as I get time but there is a lot to check.

**Comment by Starfox64** [comment]
Sorry for bumping this but is this PR still alive?

I was hoping this could be merged for or before 9.00. I'm happy to help if required.

I'm asking to know if I should wait on an official implementation or just roll out my own like many have done by overriding the library.

**Comment by skroczek** [comment]
We are now using zitadel as our IDP for various reasons. We have also switched from PHP to Go as our main language. Because of this, I unfortunately have neither much time nor interest to take care of it anymore. Nevertheless, this remains an excellent library. I will of course leave the PR open so that it can be merged with the repository if desired.

**Comment by fredsal** [comment]
Hi, is this PR likely to be merged soon?

**Comment by parallels999** [comment]
> No timescale on this I'm afraid.

@Sephster still no timescale?

**Comment by MJTheOne** [comment]
Would also like to know why this is "ignored"? Custom claims are perfectly normal

**Comment by parallels999** [comment]
@Sephster is possible merging or rebasing this with master?

**Comment by systemsolutionweb** [comment]
@Sephster hi, we would love to have this feature
any news?

This is the PR with the most comments

Thanks

File: src/Grant/AuthCodeGrant.php [hunk 14]
```diff
@@ -161,8 +161,18 @@ public function respondToAccessTokenRequest(
             }
         }
 
+        $privateClaims = [];
+
+        if ($this->claimRepository !== null) {
```
Comment (MJTheOne) [inline src/Grant/AuthCodeGrant.php:166]: First of all, sorry for the (very) late comment. But just got an email that people are interested again.
Maybe just personal preference, but checking if something is not null is not actually checking anything (because it can be ANYTHING at this point).
I'd rather check if it is what I want it to be, e.g. an instanceof check.

**Comment by ssw1cblarrion** [comment]
Any news?
